benchmarks/babelstream/src/benchmarkTest.cpp

```diff
@@ -0,0 +1,434 @@
+    {
+        const T scalar = static_cast<T>(startScalar);
+        auto const i = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0];
+        c[i] = a[i] + scalar * b[i];
+    }
+};
+
+//! Dot product of two vectors. But the result is not a scalar, a vector of block level dot products. For the
+//! BabelStream implementation and documenation: https://github.com/UoB-HPC
+struct DotKernel
+{
+    //! The kernel entry point
+    //! \tparam TAcc The accelerator environment to be executed on.
+    //! \tparam T The data type
+    //! \param acc The accelerator to be executed on.
+    //! \param a Pointer for vector a
+    //! \param b Pointer for vector b
+    //! \param sum Pointer for result vector consisting sums for each block
+    template<typename TAcc, typename T>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc, T const* a, T const* b, T* sum, alpaka::Idx<TAcc> arraySize) const
+    {
+        using Idx = alpaka::Idx<TAcc>;
+        auto& tbSum = alpaka::declareSharedVar<T[blockThreadExtentMain], __COUNTER__>(acc);
+
+        auto i = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0];
+        auto const local_i = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0];
+        auto const totalThreads = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc)[0];
+
+        T threadSum = 0;
+        for(; i < arraySize; i += totalThreads)
+            threadSum += a[i] * b[i];
+        tbSum[local_i] = threadSum;
+
+        auto const blockSize = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0];
+        for(Idx offset = blockSize / 2; offset > 0; offset /= 2)
+        {
+            alpaka::syncBlockThreads(acc);
+            if(local_i < offset)
+                tbSum[local_i] += tbSum[local_i + offset];
+        }
+
+        auto const gridBlockIndex = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0];
+        if(local_i == 0)
+            sum[gridBlockIndex] = tbSum[local_i];
+    }
+};
+
+//! \brief The Function for testing babelstream kernels for given Acc type and data type.
+//! \tparam TAcc the accelerator type
+//! \tparam DataType The data type to differentiate single or double data type based tests.
+template<typename TAcc, typename DataType>
+void testKernels()
+{
+    using Acc = TAcc;
+    // Define the index domain
+    // Set the number of dimensions as an integral constant. Set to 1 for 1D.
+    using Dim = alpaka::Dim<Acc>;
+    using Idx = alpaka::Idx<Acc>;
+
+    // Print the accelerator, data type and vector size
+    std::cout << "\nUsing alpaka accelerator: " << alpaka::getAccName<Acc>() << std::endl;
+    std::string dataTypeStr;
+    if(std::is_same<DataType, float>::value)
+    {
+        dataTypeStr = "single-precision";
+    }
+    else if(std::is_same<DataType, double>::value)
+    {
+        dataTypeStr = "double-precision";
+    }
+    std::cout << "Data type: " << typeid(DataType).name() << " " << dataTypeStr << std::endl;
```

Comment:
Do you know, if there is some reporting function to add the information as key value pair to the output? At the moment, it simply captures the stdout and write a string to the xml, which make it hard to parse.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Catch2TestRun name="babelstream" rng-seed="1921823939" xml-format-version="2" catch2-version="3.3.2">
  <TestCase name="Babelstream Bechmark by Alpaka Kernel - Single P. - TestAccs1D - 0" tags="[benchmark-test]" filename="/home/simeon/projects/alpaka/benchmarks/babelstream/src/benchmarkTest.cpp" line="418">
    <BenchmarkResults name="CopyKernel" samples="100" resamples="100000" iterations="1" clockResolution="32.5898" estimatedDuration="1.62794e+07">
      <!-- All values in nano seconds -->
      <mean value="162549" lowerBound="162235" upperBound="163277" ci="0.95"/>
      <standardDeviation value="2305.52" lowerBound="1253.62" upperBound="4617.02" ci="0.95"/>
      <outliers variance="0.0750291" lowMild="0" lowSevere="0" highMild="1" highSevere="8"/>
    </BenchmarkResults>
    <BenchmarkResults name="AddKernel" samples="100" resamples="100000" iterations="1" clockResolution="32.5898" estimatedDuration="2.07714e+07">
      <!-- All values in nano seconds -->
      <mean value="208480" lowerBound="207810" upperBound="209855" ci="0.95"/>
      <standardDeviation value="4688.39" lowerBound="2825.43" upperBound="8760.11" ci="0.95"/>
      <outliers variance="0.161071" lowMild="0" lowSevere="0" highMild="1" highSevere="6"/>
    </BenchmarkResults>
    <BenchmarkResults name="MulKernel" samples="100" resamples="100000" iterations="1" clockResolution="32.5898" estimatedDuration="1.19507e+07">
      <!-- All values in nano seconds -->
      <mean value="122044" lowerBound="121594" upperBound="122670" ci="0.95"/>
      <standardDeviation value="2676.42" lowerBound="2061.12" upperBound="3957.46" ci="0.95"/>
      <outliers variance="0.151683" lowMild="0" lowSevere="0" highMild="1" highSevere="1"/>
    </BenchmarkResults>
    <BenchmarkResults name="TriadKernel" samples="100" resamples="100000" iterations="1" clockResolution="32.5898" estimatedDuration="2.11882e+07">
      <!-- All values in nano seconds -->
      <mean value="212046" lowerBound="211493" upperBound="212948" ci="0.95"/>
      <standardDeviation value="3544.51" lowerBound="2546.6" upperBound="5707.04" ci="0.95"/>
      <outliers variance="0.094408" lowMild="0" lowSevere="0" highMild="1" highSevere="9"/>
    </BenchmarkResults>
    <OverallResult success="true" skips="0">
      <StdOut>
========================
Time:2024-07-01 14:21:45

Using alpaka accelerator: AccCpuSerial&lt;1,unsigned int>
Data type: f single-precision
Data size: 1048576 elements
Work division: {gridBlockExtent: (1048576), blockThreadExtent: (1), threadElemExtent: (1)}
      </StdOut>
    </OverallResult>
  </TestCase>
```

Reply:
Yes I will use "INFO". Good idea thanks. It turned out to be `-r json` also works for `catch2 3.5.2`. I am doing it for json, more readable.

Reply:
json reporter seem to not include benchmark results only records REQUIRE results etc. XML reporter prints benchmark results as expected.

Reply:
Only WARN has worked. I will use `WARN`.  
```

WARN(dataType);  // This will be included as custom XML in the report
    INFO("\n normal info Data type is" << dataType);
    UNSCOPED_INFO("unscoped Data type is" << dataType);
    INFO("info example");
    UNSCOPED_INFO("unscoped example");
    INFO("normal info array size is" << arraySize);
    UNSCOPED_INFO("unscoped array size is" << arraySize);
    INFO("normal info data size is" << dataSize);
    UNSCOPED_INFO("unscoped data size is" << dataSize);
    CAPTURE(dataSize);  // This will be included as custom XML in the report
    CAPTURE(dataType);
   SUCCEED("SUCCEED");  // This will be included as custom XML in the report
```

Reply:
This will need additional operation to extract data from fields. May be I should use cout output directly and convert it to json by a script instead of dealing with this...
```
   <Warning filename="/home/yusufo81/projects/alpaka-dir/alpaka/benchmarks/babelstream/src/benchmarkTest.cpp" line="146">
      TimeStamp:2024-07-09 15:06:59
    </Warning>
    <Warning filename="/home/yusufo81/projects/alpaka-dir/alpaka/benchmarks/babelstream/src/benchmarkTest.cpp" line="147">
      DataSize:1048576
    </Warning>
    <Warning filename="/home/yusufo81/projects/alpaka-dir/alpaka/benchmarks/babelstream/src/benchmarkTest.cpp" line="148">
      DataSize:single-precision
    </Warning>
```